telemetry/metrics.go

```diff
@@ -70,61 +33,85 @@ type GatherResponse struct {
 	ContentType string
 }
 
-// New creates a new instance of Metrics
-func New(cfg Config) (_ *Metrics, rerr error) {
-	if !cfg.Enabled {
+// New creates a new instance of metrics
+func New(opts ...Option) (Metrics, error) {
+	var err error
+
+	c := new(Config)
+	c.loadDefaults()
+	for _, fn := range opts {
```

Comment:
I don't understand this opts and config juggling. Really only a `Config` should be needed. I can't see a strong argument for the need of `Opts*`.

E.g.

```go
cfg := new(Config)
cfg.LoadDefaults() // this should be public

// cfg.Foo = ...
```

Typically, the entire `Opts*` and `With` design works when you're modifying a type directly, e.g. `BaseApp`.

Just seems odd to me. @kocubinski wdyt?

Reply:
I agree that it seems odd to configure using both a struct (`Config`) and an options functor pattern, we should pick just one.  Was there a reason to choose both @harpy-wings ?

Reply:
Using "Options" pattern provides more flexibility and extendability for the codebase, and will lead to a cleaner codebase.
The decision is yours whether to follow this pattern or not, but if you want to migrate to this approach, this can be done over a big refactor or by some small PR that supports both approaches, after a while, you can make the struct pattern deprecated.

Reply:
struct pattern is just setting some values and sometimes based on some value, you should do some more processes, in the struct pattern, all these processes should go to the `New` functions and you will have a lot of `if/else` conditions in the construction function, but in the Options pattern each process goes to the relative function, you can put some validation over there and even you can have the factory of factory pattern for the Options which gives us many advantages in the SDK projects.

Reply:
I'm interested in the conversation, if you want we can have a discussion meeting about this.

Reply:
I also generally prefer the Options pattern since it's more flexible, is it worth the API break though @alexanderbez?  If we do this `Config` type should not be exported.

Another option is (as you said) is to provide a `NewWithOptions` constructor (non API breaking) with plenty of docs so that people aren't confused about which to use.

Reply:
Absent specific rationale, there's no reason to refactor a config-struct-based API over to functional options. Neither pattern is better or worse than the other.

Reply:
resolved.